python/lsst/daf/butler/_formatter.py

```diff
@@ -51,12 +60,927 @@
     from ._dataset_ref import DatasetRef
     from ._dataset_type import DatasetType
     from ._storage_class import StorageClass
-    from .dimensions import DataCoordinate
+    from .datastore.cache_manager import AbstractDatastoreCacheManager
 
     # Define a new special type for functions that take "entity"
     Entity: TypeAlias = DatasetType | DatasetRef | StorageClass | str
 
 
+class FileIntegrityError(RuntimeError):
+    """The file metadata is inconsistent with the metadata supplied by
+    the datastore.
+    """
+
+
+class FormatterV2(metaclass=ABCMeta):
+    """Interface for reading and writing datasets using URIs.
```

Comment:
One thing I was a little uncomfortable with in the original `Formatter` interface is that the same class is responsible for both reading and writing, but the state the class wants to hold for reading is not generally the same as the state it wants to read for writing.  It's nice to have one class name to configure rather than two, but other than that I think reading and writing are best kept separate.

So, while I realize this would be another pretty big change (apologies for not raising the possibility earlier), what do you think about making the `write` interface go entirely through `classmethod`s, so that constructing an instance is something you do only for reads?  The reason I'd let the instance state be for reads is to give the formatter the URL and `DatasetRef` construction to let it do some work up front, and then support multiple component or parameter reads that it may be able to optimize internally (e.g. using [fsspec's buffering/caching file-like object wrappers](https://filesystem-spec.readthedocs.io/en/latest/api.html#readbuffering)).  Getting multiple components or sets of parameters would require new interfaces higher up, too, of course, but I think that's easier than the formatter side.

If we don't go this route, I think we want to be very clear about these being two-state instances that are constructed in different ways for read and write and may have different attributes set as a result.  And we really want to limit to just two states.

Reply:
Writing currently needs to be an instance method because the write parameters are read from configuration and attached to the formatter instance by the formatter factory.

I've only just read the second half of your second paragraph where you are suggesting that `formatter.read()` might want to be called multiple times with different components. Datastore isn't set up to be able to do that at the moment because butler only asks for one component for one dataset or else all components from a disassembled dataset.

Reply:
I do think reading N components might be useful, but I don't have a concrete use case in mind and I could imagine it being better implemented by being able to pass a list of components rather than by reworking what is considered formatter instance state.  It's not something I'd go add at a high level right now, just something to keep in mind as a possibility (e.g. is the formatter interface such that we could add a method for that with a default implementation and override it in just some concrete formatters, rather than having to change them all again).

Reply:
Now that the user subclasses something like `read_from_uri(uri, component)` and doesn't subclass `read` itself then we could pass multiple components to `read` and have that by default be N calls to `read_from_uri`.